File: server/api/practice.ts
import { Router, Request, Response } from "express";
import { PracticeSessionRequest, ErrorBatchRequest } from "../types.js";
import { createPracticeSession, createErrorBatch, PracticeError } from "../services/practice-service.js";

const router = Router();

// POST /api/practice/session
router.post("/practice/session", (req: Request, res: Response) => {
  const parsed = PracticeSessionRequest.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ detail: parsed.error.errors });
    return;
  }
  try {
    const result = createPracticeSession(parsed.data);
    res.json(result);
  } catch (exc) {
    if (exc instanceof PracticeError) {
      res.status(400).json({ detail: exc.message });
    } else {
      res.status(500).json({ detail: "Internal error" });
    }
  }
});

// POST /api/errors/batch
router.post("/errors/batch", (req: Request, res: Response) => {
  const parsed = ErrorBatchRequest.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ detail: parsed.error.errors });
    return;
  }
  try {
    const result = createErrorBatch(parsed.data);
    res.json(result);
  } catch (exc) {
    if (exc instanceof PracticeError) {
      res.status(400).json({ detail: exc.message });
    } else {
      res.status(500).json({ detail: "Internal error" });
    }
  }
});

export default router;

File: backend/api/practice.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from backend.db.database import get_db
from backend.schemas import ErrorBatchRequest, ErrorBatchResponse, PracticeSessionRequest, PracticeSessionResponse
from backend.services import practice_service

router = APIRouter()


@router.post("/practice/session", response_model=PracticeSessionResponse)
def create_practice_session(payload: PracticeSessionRequest, db: Session = Depends(get_db)) -> PracticeSessionResponse:
    try:
        return practice_service.create_practice_session(db, payload)
    except practice_service.PracticeError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post("/errors/batch", response_model=ErrorBatchResponse)
def create_error_batch(payload: ErrorBatchRequest, db: Session = Depends(get_db)) -> ErrorBatchResponse:
    try:
        return practice_service.create_error_batch(db, payload)
    except practice_service.PracticeError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
